stackhpc_ipa_hardware_managers/system_nic.py

```diff
@@ -0,0 +1,367 @@
+              nic_firmware='[{"vendor_id": "15B3", "device_id": "1013", \\
+              "firmware_version": "12.20.1010"}]'
+
+        Each network card is checked against each of the items in the matching
+        criteria list. If the device and vendor ids match, the firmware version
+        is checked against the expected version specified in the
+        nic_firmware property's firmware_version field.
+
+        """
+
+        if _is_nic_verification_disabled(node):
+            LOG.warning('NIC firmware version verification has been disabled.')
+            return {}
+
+        if "extra" not in node or "nic_firmware" not in node["extra"]:
+            raise errors.CleaningError(
+                "Expected property 'nic_firmware' not found. " +
+                _HELP_MSG_EXAMPLE)
+
+        firmware_matchers = node["extra"]["nic_firmware"]
+
+        if not isinstance(firmware_matchers, list):
+            raise errors.CleaningError(
+                "The property 'nic_firmware' should be a list. " +
+                _HELP_MSG_EXAMPLE
+            )
+
+        lookup_table = _get_pci_id_lookup_table(firmware_matchers)
+        successes, failures = self._process_expected_versions(lookup_table)
+
+        error_msgs = []
+        for interface, result in failures.iteritems():
+            msg = (
+                "Firmware version mismatch for card: {interface}. The "
+                "expected version was: {expected_version}, "
+                "but the actual version was {actual_version}. "
+                "The matcher that failed was {matcher}"
+            ).format(
+                interface=interface,
+                expected_version=result.expected_version,
+                actual_version=result.actual_version,
+                matcher=result.matcher)
+            error_msgs.append(msg)
+
+        if failures:
+            raise errors.CleaningError(
+                "Found {} firmware version mismatches "
+                "when verifying NIC firmware. The errors were: \n"
+                .format(len(error_msgs)) +
+                "\n".join(error_msgs)
+            )
+
+        return successes
+
+    def _process_expected_versions(self, pci_matcher_lookup_table):
+        """Processes all matching rules
+
+        where a matching rule is a dict with the following required fields:
+
+        * vendor_id : a numeric pci vendor id
+        * device_id: a numeric pci device id
+        * firmware_version: expected firmware version
+
+        :param pci_matcher_lookup_table: dictionary mapping pci_id to a
+            matching rule
+        :return: tuple with the structure (successes, failures) where,
+            *successes* and *failures* are dictionaries mapping the
+            interface name to a NICFirmwareVerifyResult
+        """
+
+        interfaces = self.get_interface_descriptors()
+        successes = {}
+        failures = {}
+        assert type(self.get_interface_descriptors()) == list
```

Comment:
Is this left over debug?

Reply:
Would you remove the assert, or stop it duplicating the call and leave it in?

Reply:
I don't mind too much, it just stood out as the only assert in the code, checking something quite fundamental to a method in the same class. I guess the duplicate method call was a bit odd too.